### DISCUSSION TITLE
自定义控件中的TemplatePart的样式问题

### DISCUSSION THREAD

**Discussion by yangf85** [BYJRK/DotNet-Discussions#95] [WPF 开发]
大神们，比如现在有一个文件路径选择控件，使用了TextBox和Button来作为TemplatePart,给TextBox和Button暴露出两个Style类型的依赖属性供外部使用，现在存在一个问题，在文件路径选择控件中设置了这两个控件的样式后，在有多个文件路径选择控件的情况下，这些样式只会应用到最后一个控件上，必须把样式资源设置为IsShared=False,这是什么原因？

**Comment by BYJRK** [comment]
在WPF中，每个控件都有一个默认的样式资源，这些资源在控件被创建时会被初始化并应用。当你为一个控件设置了样式时，该样式会成为该控件的自定义样式，并会覆盖默认样式。在没有设置`IsShared`属性的情况下，默认情况下样式资源是共享的，这意味着多个控件可以共享同一个样式资源。

问题出在`IsShared`属性上。当你将样式资源的`IsShared`属性设置为`False`时，每个控件会创建自己的样式实例，而不再共享样式资源。这意味着每个文件路径选择控件都有自己独立的样式，并且样式的修改不会影响其他控件。

如果你没有设置`IsShared`属性或将其设置为`True`，那么样式资源将被视为共享资源，因此在多个文件路径选择控件中使用相同的样式时，它们会共享同一个样式实例。这导致最后一个控件上设置的样式会覆盖之前的样式，因为它们实际上引用的是同一个样式实例。

通过将`IsShared`属性设置为`False`，你确保每个控件都有自己独立的样式实例，这样每个控件可以根据自己的样式设置来呈现，而不会互相干扰。这样就可以解决你描述的问题。

（by ChatGPT）
